### ISSUE TITLE
Dependencies in filters - strange situation

### ISSUE THREAD

**Issue by ArekZw** [o10n-x/wordpress-javascript-optimization#24]
js.async.filter.config:
Case 1:
```
          {
            "match": "jquery",
            "async": true,
            "minify": true
          },
          {
            "match": "wpmm.js",
            "async": true,
            "dependency": "jquery"
          },
          {
            "match": "sticky",
            "async": true,
            "dependency": "jquery"
          },

```
Case 2:
```
          {
            "match": "jquery",
            "async": true,
            "minify": true
          },
          {
            "match": "wpmm.js",
            "async": true,
            "dependency": "sticky"
          },
          {
            "match": "sticky",
            "async": true,
            "dependency": "jquery"
          },
```

Case 1:
jquery - loaded
wpmm - NOT loaded
sticky - loaded 

Case 2:
jquery - loaded 
wpmm - loaded
sticky - loaded

Why?! 
I completely do not understand why the same dependency works for one file and not the other (case 1). Match for wpmm is correct (the 2nd case is proof).

In code all looks good:
Case 1:
```
[1,"https:\/\/mywebsite.com\/wp-content\/jquery.js",null,null,null,null]
[1,"https:\/\/mywebsite.com\/wp-content\/sticky.min.js",null,null,null,"jquery"],
[1,"https:\/\/mywebsite.com\/wp-content\/wpmm.js",null,null,null,"jquery"]

```
```
[1,"https:\/\/mywebsite.com\/wp-content\/jquery.js",null,null,null,null],
[1,"https:\/\/mywebsite.com\/wp-content\/sticky.min.js",null,null,null,"jquery"],
[1,"https:\/\/mywebsite.com\/wp-content\/wpmm.js",null,null,null,"sticky"]
```

**Comment by optimalisatie** [comment]
Hi @ArekZw,

It was a bug in the dependency loader that has been fixed in `v0.0.86`.

Thank you very much for reporting the bug with such detailed feedback! We didn't discover it before since we normally concatenate scripts and use just 1 dependency.

We've added a new option in `v0.0.87` to match dependencies using a regular expression. 

```json
{
	"match": "script",
	"async": true,
	"dependency": [
		"dependency-a",
		{
		  "match": "/dependency-(b|c)/",
		  "regex": true
		}
	]
}
```

Tip: When selecting jQuery as dependency, make sure that you match the specific `jquery.js` script. Many jQuery modules contain the word `jquery`.

![image](https://user-images.githubusercontent.com/8843669/49580096-6ca02500-f94e-11e8-90b0-cc374badd9c3.png)

![image](https://user-images.githubusercontent.com/8843669/49578307-bf2b1280-f949-11e8-8f42-7e266ce8c15e.png)

**Comment by ArekZw** [comment]
Hmm... Now I have error:
```
(index):4 Uncaught TypeError: c.querySelectorAll is not a function
    at aa ((index):4)
    at (index):22
    at H ((index):11)
    at (index):22
    at (index):27
```
Even if my config is only:
js.async.enabled => true,

**Comment by optimalisatie** [comment]
Thank you for the continued feedback despite the issues!

It seems that for some reason the inline script did not retrieve the `document.head` element. When `document.head` is undefined/null it will use `document.getElementsByTagName('head')[0]` as fallback. For some reason, in your case, that method also returned null.

We've made the retrieval of the head element more robust in the latest update. It will now display a an error when using debug mode and it will fallback to `document.documentElement` to perform a query when the `<head>` element is missing from the HTML.

Could you please test the following code in the browser console:

```javascript
console.log('document.head',document.head);
console.log('querySelectorAll',typeof document.querySelectorAll);
console.log('querySelectorAll on <head>',typeof document.head.querySelectorAll);
```

**Comment by ArekZw** [comment]
![zrzut ekranu z 2018-12-06 20-10-06](https://user-images.githubusercontent.com/11788346/49606240-1d301800-f993-11e8-9313-8166339dc97c.png)

**Comment by ArekZw** [comment]
Indeed, I had 0.87, I did not notice 0.0.88.
Now I updated to 0.0.89 and...
I still have aa() function in code. It look like this inline script is somewhere cached..

Even if I remove core/public/js/o10n.js, this old code with aa() is still in source code. :|
On dev server and on localhost. I'm sure that plugin is updated. 

I remove o10n cache folder in wp-content. I restart nginx. I not have other cache. I searched the database for this script and nothing found... This old inline script is still in code aggghhh...

I will look for a cause tomorrow. Thank you for help.
